### PR TITLE
PullRequest for Issue2248: code refactor for AppControllers for generalization

### PR DESCRIPTION
**Pull request by XuetaoLiu-CLS**
- refine the startup() procedure of CLSAppController
  . removed makeConnection()
  . create setupDataFolder() template call
  . add setupScanConfiguration() function
  . create template procedure for setupUserInterface() and related template functions

- implement the template function for each beamlines and move codes around to fit the template

### CODE REVIEW

**Comment by XuetaoLiu-CLS** [comment]
Tested on SXRMB. The statup and configuration are fine. The EXAFS scan won't start -- stuck at the first energy.

**Comment by XuetaoLiu-CLS** [comment]
Tested on BioXAS Side, REIXAS, IDEAS, VESPERS, SXRMB.

**Comment by XuetaoLiu-CLS** [comment]
@helfrij can you review this?

I was simply shifting codes around instead of changing the functionality. As a result, +1,027 codes and −1,127 codes :8ball: .

You might want to look at CLSAppController first, which will give you a quick idea on what I was trying to do.

Thanks.

**Comment by helfrij** [comment]
Looks good, @XuetaoLiu-CLS. Nice work!

Time Estimate: 30 minutes.

File: source/application/SXRMB/SXRMBAppController.cpp [hunk 1]
```diff
@@ -459,18 +407,14 @@ void SXRMBAppController::onCurrentScanActionFinishedImplementation(AMScanAction
...
-		exafsScanConfiguration_->addRegionOfInterest(region);
-		microProbe2DOxidationScanConfiguration_->addRegionOfInterest(region);
-	}
-
 	// This is connected here because we want to listen to the detectors for updates, but don't want to double add regions on startup.
 	connect(detector, SIGNAL(addedRegionOfInterest(AMRegionOfInterest*)), this, SLOT(onRegionOfInterestAdded(AMRegionOfInterest*)));
 	connect(detector, SIGNAL(removedRegionOfInterest(AMRegionOfInterest*)), this, SLOT(onRegionOfInterestRemoved(AMRegionOfInterest*)));
 	connect(detector, SIGNAL(regionOfInterestBoundingRangeChanged(AMRegionOfInterest*)), this, SLOT(onRegionOfInterestBoundingRangeChanged(AMRegionOfInterest*)));
+
+	foreach (AMRegionOfInterest *region, userConfiguration_->regionsOfInterest()){
```
Comment (dretrex) [inline source/application/SXRMB/SXRMBAppController.cpp]: Maybe I missed it somewhere, but where do you set the regions of interest to the scan configurations?

**Comment by dretrex** [comment]
Overall, this looks good. Just to double check, did you run any scans when doing the testing for this?  The main thing I found hard to follow was whether all the proper things were getting connected.
Time Estimate: 10 minutes

File: source/application/SXRMB/SXRMBAppController.cpp (hunk 1 above)
Comment (XuetaoLiu-CLS) [inline source/application/SXRMB/SXRMBAppController.cpp]: This is the slot for signal `loadedFromDb()`. When we load the user configuration from the db in function `setupUserConfiguration()`, it will emit the signal. 

Therefore, the ROIs of the user configuration are created when we execute  `userConfiguration_->loadFromDb(AMDatabase::database("user"), 1)`.

Comment (XuetaoLiu-CLS) [inline source/application/SXRMB/SXRMBAppController.cpp]: oh... I misunderstood your question here.

function `detector->addRegionOfInterest(region->createCopy());` will emit signal `emit addedRegionOfInterest(newRegionOfInterest);`.
SXRMBAppController is listening to that signal (Ln411):
`connect(detector, SIGNAL(addedRegionOfInterest(AMRegionOfInterest*)), this, SLOT(onRegionOfInterestAdded(AMRegionOfInterest*)));`

and in slot `onRegionOfInterestAdded`, we added the ROIs to the userConfiguration, and scanConfiguraitons.

**Comment by XuetaoLiu-CLS** [comment]
@dretrex I didn't run a scan when testing this. I also didn't test it on SGM.

Maybe I should retest it Monday on all the beamliens.

**Comment by davidChevrier** [comment]
@dretrex do you want to help @XuetaoLiu-CLS coordinate some testing?
There's no need to test anything on SGM, they've explicitly asked for no updates to be made going forward.

**Comment by dretrex** [comment]
I'm going to be on VESPERS and IDEAS next week anyway.  I can run some test scans on those beamlines.  Maybe @helfrij can run a test on BioXAS as well?

**Comment by helfrij** [comment]
Yup, no problem.

**Comment by XuetaoLiu-CLS** [comment]
@davidChevrier @dretrex @helfrij Thanks :+1: 

I will test on SXRMB

**Comment by XuetaoLiu-CLS** [comment]
@helfrij tested on BioXAS. The scans work as expected and I made a changes according to her testing result.
